Adopt API for process launch and capabilities

pvollan opened the pull request:
#### 6cb85cc698675e4f4141b72197223b3f0fa0cb15
<pre>
Adopt API for process launch and capabilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=268644">https://bugs.webkit.org/show_bug.cgi?id=268644</a>
<a href="https://rdar.apple.com/122196452">rdar://122196452</a>

Reviewed by Brent Fulgham and Chris Dumez.

Adopt BrowserEngineKit API for process launch and capabilities.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setIsHoldingLockedFiles):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::acquireLockedFileActivity):
(WebKit::NetworkProcess::invalidateActivity):
(WebKit::NetworkProcess::hasAcquiredActivity const):
(WebKit::NetworkProcess::acquireLockedFileGrant): Deleted.
(WebKit::NetworkProcess::invalidateGrant): Deleted.
(WebKit::NetworkProcess::hasAcquiredGrant const): Deleted.
* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::platformCapability const):
* Source/WebKit/Platform/cocoa/ExtensionCapability.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ExtensionCapabilityGrant::ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::setPlatformGrant):
* Source/WebKit/Platform/cocoa/MediaCapability.h:
* Source/WebKit/Platform/cocoa/MediaCapability.mm:
(WebKit::createPlatformCapability):
(WebKit::MediaCapability::environmentIdentifier const):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::extensionProcess const):
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::grantCapability):
(WebKit::grantCapabilityInternal):
(WebKit::ExtensionCapabilityGranter::setMediaCapabilityActive):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::gpuProcessCachesDirectory):
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::extensionProcess const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::launchWithExtensionKit):
(WebKit::invalidateProcess):
(WebKit::ProcessLauncher::platformDestroy):
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::platformInvalidate):
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]):
(-[WKContentView _setupVisibilityPropagationViewForGPUProcess]):

Canonical link: <a href="https://commits.webkit.org/274822@main">https://commits.webkit.org/274822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d05700de2a2314a02f91e0fe8f4613649935db9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39230 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41583 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41764 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35130 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15538 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32828 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39804 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15331 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34011 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13312 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13290 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43042 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35626 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35280 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39095 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11595 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37330 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15648 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15311 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5298 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15134 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -37,10 +38,10 @@
-static RetainPtr<_SECapability> createPlatformCapability(const URL& url)
+static RetainPtr<BEProcessCapability> createPlatformCapability(const URL& url, RetainPtr<BEMediaEnvironment> mediaEnvironment)
 {
 #if USE(EXTENSIONKIT)
```
Do we have cases of `EXTENSION_CAPABILITIES` without `EXTENSIONKIT`? This seems like Cocoa-only code, so we might be able to reduce this macro complexity slightly.

pvollan replied:
Ah, good point, will fix!

brentfulgham commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -50,7 +51,8 @@
 MediaCapability::MediaCapability(URL url)
     : m_url { WTFMove(url) }
-    , m_platformCapability { createPlatformCapability(m_url) }
+    , m_mediaEnvironment(adoptNS([[BEMediaEnvironment alloc] initWithWebPageURL:url]))
```
`url` has moved at this point -- seems like you should be using `m_url` here.

pvollan replied:
Ah, good catch, will fix!

Thanks for reviewing!

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -255,10 +264,17 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+        if (error)
+            RELEASE_LOG_ERROR(ProcessCapabilities, "%{public}s failed with error: %{public}@", __FUNCTION__, error);
+        else
+          return ExtensionCapabilityActivationPromise::createAndResolve();
```
Nit: Whitespace is a little off on this line!

brentfulgham reviewed (approved):
I had a few small suggestions, but overall looks good. r=me

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.h:
```diff
@@ -64,12 +66,15 @@ class MediaCapability final : public ExtensionCapability, public CanMakeWeakPtr<
     // ExtensionCapability
     String environmentIdentifier() const final;
-    RetainPtr<_SECapability> platformCapability() const final { return m_platformCapability.get(); }
+    RetainPtr<BEProcessCapability> platformCapability() const final { return m_platformCapability.get(); }
```
Plenty of functions in this PR return smart pointers when they do not need to.

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -37,10 +38,10 @@
 using WebCore::RegistrableDomain;
 
-static RetainPtr<_SECapability> createPlatformCapability(const URL& url)
+static RetainPtr<BEProcessCapability> createPlatformCapability(const URL& url, RetainPtr<BEMediaEnvironment> mediaEnvironment)
```
Passing a smart pointer in parameter by value is not great.

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -51,24 +53,31 @@
     NSError *error = nil;
-    _SEGrant *grant = [process grantCapability:capability error:&error];
+    BEProcessCapabilityGrant* grant = nil;
+    if (std::holds_alternative<RetainPtr<BEWebContentProcess>>(process))
```
Can we use WTF::switchOn()?

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -372,7 +372,7 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
 {
 #if USE(EXTENSIONKIT)
-    if (process.extensionProcess()) {
+    if (process.extensionProcess().index() != std::variant_npos) {
```
May be nicer to use `valueless_by_exception()`

cdumez replied:
Actually, see my comment above about calling `ExtensionProcess()` I do not believe this does anything at the moment?

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -465,9 +465,15 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
     RELEASE_LOG(ProcessSuspension, "%p - ProcessAssertion::acquireSync Trying to take RBS assertion '%{public}s' for process with PID=%d", this, m_reason.utf8().data(), m_pid);
 #if USE(EXTENSIONKIT)
-    if (m_process && m_capability) {
+    if (m_process.index() != std::variant_npos && m_capability) {
```
May be nicer to use `valueless_by_exception()`

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -465,9 +465,15 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
         NSError *error = nil;
-        m_grant = [m_process grantCapability:m_capability->platformCapability().get() error:&error];
+        auto capability = m_capability->platformCapability().get();
+        if (std::holds_alternative<RetainPtr<BEWebContentProcess>>(m_process))
```
`WTF::switchOn()`?

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -121,7 +124,7 @@ class ProcessLauncher : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<P
 
 #if USE(EXTENSIONKIT)
-    RetainPtr<_SEExtensionProcess> extensionProcess() const { return m_process; }
+    ExtensionProcess extensionProcess() const { return m_process; }
```
Should probably return a `const ExtensionProcess&` to avoid churn.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +85,30 @@
 }
 
-static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLauncher::ProcessType processType, ProcessLauncher::Client* client, std::function<void(ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, _SEExtensionProcess* process, ASCIILiteral name, NSError* error)> handler)
+static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLauncher::ProcessType processType, ProcessLauncher::Client* client, std::function<void(ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, ExtensionProcess process, ASCIILiteral name, NSError* error)> handler)
```
Passing `ExtensionProcess` by value is not great given that it contains smart pointers. It's causing refcounting churn.

cdumez replied:
Still being passed by value.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -139,10 +137,22 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+#if USE(EXTENSIONKIT)
+static void invalidateProcess(ExtensionProcess process)
+{
+    if (std::holds_alternative<RetainPtr<BEWebContentProcess>>(process))
```
WTF::switchOn()?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -181,21 +191,29 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
-            OSObjectPtr xpcConnection = [process makeLibXPCConnectionError:&error];
+            OSObjectPtr<xpc_connection_t> xpcConnection;
+
+            if (std::holds_alternative<RetainPtr<BEWebContentProcess>>(process))
```
WTF::switchOn()?

cdumez commented on Source/WebKit/UIProcess/ProcessAssertion.h:
```diff
@@ -72,7 +75,7 @@ class ProcessAssertion : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<
     enum class Mode : bool { Sync, Async };
 #if USE(EXTENSIONKIT)
-    static Ref<ProcessAssertion> create(RetainPtr<_SEExtensionProcess>, const String& reason, ProcessAssertionType, Mode = Mode::Async, const String& environmentIdentifier = emptyString(), CompletionHandler<void()>&& acquisisionHandler = nullptr);
+    static Ref<ProcessAssertion> create(ExtensionProcess, const String& reason, ProcessAssertionType, Mode = Mode::Async, const String& environmentIdentifier = emptyString(), CompletionHandler<void()>&& acquisisionHandler = nullptr);
```
We shouldn't pass `ExtensionProcess` by value.

cdumez commented on Source/WebKit/UIProcess/ProcessAssertion.h:
```diff
@@ -47,8 +47,11 @@ OBJC_CLASS WKRBSAssertionDelegate;
+OBJC_CLASS BENetworkingProcess;
+OBJC_CLASS BERenderingProcess;
+OBJC_PROTOCOL(BEProcessCapabilityGrant);
+using ExtensionProcess = std::variant<RetainPtr<BEWebContentProcess>, RetainPtr<BENetworkingProcess>, RetainPtr<BERenderingProcess>>;
```
Note that given all the cases where you need to switchOn() this, it might look nicer to have a class to wrap this std::variant<> and abstract away all this complexity. e.g. this class would have an "invalidate()" member function which would do the `WTF::switchOn()` internally. Call sites would look a lot nicer I think.

cdumez commented on Source/WebKit/Platform/cocoa/AssertionCapability.mm:
```diff
@@ -47,7 +47,7 @@
 }
 
-RetainPtr<_SECapability> AssertionCapability::platformCapability() const
+RetainPtr<BEProcessCapability> AssertionCapability::platformCapability() const
```
Doesn't need to return a smart pointer.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -28,21 +28,21 @@
 namespace WebKit {
 
-static void platformInvalidate(RetainPtr<_SEGrant>&& platformGrant)
+static void platformInvalidate(RetainPtr<BEProcessCapabilityGrant>&& platformGrant)
```
Why is this taking a `RetainPtr<>&&`, we don't seem to adopt the grant or transfer ownership in any way. We can probably pass a `BEProcessCapabilityGrant *`

pvollan replied:
Would you be OK to keep these for now, since they are virtual functions and will require further changes?

Thanks for reviewing?

pvollan replied:
Actually, the changes aren't that many, so I'll go ahead and make that change :)

cdumez replied:
You didn't make that change?

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.h:
```diff
@@ -64,12 +66,15 @@ class MediaCapability final : public ExtensionCapability, public CanMakeWeakPtr<
-    RetainPtr<_SECapability> platformCapability() const final { return m_platformCapability.get(); }
+    RetainPtr<BEProcessCapability> platformCapability() const final { return m_platformCapability.get(); }
+
+    RetainPtr<BEMediaEnvironment> platformMediaEnvironment() const { return m_mediaEnvironment; }
```
Doesn't need to return a smart pointer and cause refcounting churn.

pvollan replied:
Will fix!

cdumez replied:
You didn't make that change?

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -37,10 +38,10 @@
 {
 #if USE(EXTENSIONKIT)
-    if (_SECapability *capability = [get_SECapabilityClass() mediaWithWebsite:RegistrableDomain(url).string()])
+    if (auto capability = [BEProcessCapability mediaPlaybackAndCaptureWithEnvironment:mediaEnvironment.get()])
```
`auto*`?

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -77,8 +79,9 @@
 #if USE(EXTENSIONKIT)
-    if (NSString *mediaEnvironment = [m_platformCapability mediaEnvironment])
-        return mediaEnvironment;
+    auto xpcObject = [m_mediaEnvironment createXPCRepresentation];
```
Do we need to adopt here or not? Usually when there is "create" in the name, we adopt. But here it ends up calling into Swift to 🤷🏻

pvollan replied:
I do not believe we have to adopt here, even though the name suggests otherwise.

cdumez commented on Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
```diff
@@ -75,10 +75,10 @@
 {
     if (!m_processLauncher)
-        return nullptr;
+        return ExtensionProcess();
```
FYI, I believe this calls the std::variant's default constructor, which creates a value with the variant's first alternative initialized. As a result, I believe it returns a std::variant containing a `RetainPtr<BEWebContentProcess>` (because that's the first alternative) which is `nullptr`.

As a result, I doubt your checks that use `variant_npos` (or `valueless_by_exception()`) do anything here?

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -45,7 +45,10 @@
+OBJC_CLASS BEWebContentProcess;
+OBJC_CLASS BENetworkingProcess;
+OBJC_CLASS BERenderingProcess;
+using ExtensionProcess = std::variant<RetainPtr<BEWebContentProcess>, RetainPtr<BENetworkingProcess>, RetainPtr<BERenderingProcess>>;
```
To represent an "invalid" / "empty" ExtensionProcess, you could:
1. use `std::optional<ExtensionProcess>`
2. Add a `std::monostate` as first alternative: `std::variant<std::monostate, RetainPtr<BEWebContentProcess>, RetainPtr<BENetworkingProcess>, RetainPtr<BERenderingProcess>>`
3. Have your member functions return a `ExtensionProcess*` and nullptr in case of error.

cdumez reviewed (changes requested):
r- given the std::variant usage seems incorrect.

aestes commented on Source/WebKit/Platform/cocoa/AssertionCapability.h:
```diff
@@ -43,15 +43,15 @@ class AssertionCapability final : public ExtensionCapability {
     BlockPtr<void()> m_willInvalidateBlock;
     BlockPtr<void()> m_didInvalidateBlock;
-    RetainPtr<_SECapability> m_capability;
+    RetainPtr<BEProcessCapability> m_capability;
```
Since both `AssertionCapability` and `MediaCapability` now have a stored `m_capability`, we could just move this to the base class (and because `AssertionCapability ::platformCapability()` now just returns a member variable, it longer needs to return a `RetainPtr` as Chris points out).

aestes commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -51,24 +53,31 @@
+    else if (std::holds_alternative<RetainPtr<BENetworkingProcess>>(process))
+        grant = [std::get<RetainPtr<BENetworkingProcess>>(process) grantCapability:capability error:&error];
+    else if (std::holds_alternative<RetainPtr<BERenderingProcess>>(process))
+        grant = [std::get<RetainPtr<BERenderingProcess>>(process) grantCapability:capability error:&error];
```
Not relevant to this PR, but it really seems like an API design mistake that there isn't a common base class for these types where we could declare `-grantCapability:error:`

aestes commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -28,21 +28,21 @@
 
 #if ENABLE(EXTENSION_CAPABILITIES)
 
+#import <BrowserEngineKit/BECapability.h>
```
Nit: this should go below the quoted imports. Also, we should import the umbrella header.

aestes commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -28,6 +28,8 @@
 
 #if ENABLE(EXTENSION_CAPABILITIES)
 
+#import <BrowserEngineKit/BECapability.h>
```
This should go below the quoted imports. Also, we should import the umbrella header.

aestes commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -51,24 +53,31 @@
 
     NSError *error = nil;
-    _SEGrant *grant = [process grantCapability:capability error:&error];
+    BEProcessCapabilityGrant* grant = nil;
```
Nit: space goes before asterisk.

pvollan replied:
Thanks for reviewing! I will address these issues in the upcoming patch.

pvollan commented on Source/WebKit/UIProcess/ios/WKVisibilityPropagationView.mm:
```diff
@@ -47,7 +47,8 @@ - (void)propagateVisibilityToProcess:(WebKit::AuxiliaryProcessProxy&)process
         return;
 
-    RetainPtr extensionProcess = process.extensionProcess();
+#if 0
```
This is an unintended change, will fix!

cdumez commented on Source/WebKit/Platform/cocoa/AssertionCapability.mm:
```diff
@@ -39,17 +40,13 @@
-RetainPtr<_SECapability> AssertionCapability::platformCapability() const
-{
-    return m_capability;
+    RELEASE_LOG(Process, "Taking assertion %s", m_name.utf8().data());
```
You want `%{public}s` or it won't show in sysdiagnoses.

cdumez replied:
Also, you may want to add some context at the beginning of the log line (we usually include the `class::function:` that's doing the logging.

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -37,21 +38,11 @@
     : m_url { WTFMove(url) }
-    , m_platformCapability { createPlatformCapability(m_url) }
 {
+    m_mediaEnvironment = adoptNS([[BEMediaEnvironment alloc] initWithWebPageURL:m_url]);
```
Why isn't this part of the initializer list?

pvollan replied:
I did not add it to the initializer list, since it caused a style error. Would you prefer moving it into the initializer list?

cdumez replied:
Yes, this should be in the initializer list.

pvollan replied:
I agree, will move!

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -37,21 +38,11 @@
-    , m_platformCapability { createPlatformCapability(m_url) }
 {
+    m_mediaEnvironment = adoptNS([[BEMediaEnvironment alloc] initWithWebPageURL:m_url]);
+    m_platformCapability = [BEProcessCapability mediaPlaybackAndCaptureWithEnvironment:m_mediaEnvironment.get()];
```
ditto.

pvollan replied:
I added this to the function body, since `m_platformCapability` is a member of the base class.

Thanks for reviewing!

cdumez replied:
This data member should have been private in the base class. Then this constructor would either pass `platformCapability` to its base class constructor (ideally) or call a public setter on the base class (not as efficient but still OK).

cdumez commented on Source/WebKit/Platform/cocoa/MediaCapability.mm:
```diff
@@ -77,8 +68,11 @@
+    auto xpcObject = [m_mediaEnvironment createXPCRepresentation];
+    if (!xpcObject)
+        return emptyString();
+    auto identifier = xpc_dictionary_get_string(xpcObject, "identifier");
```
Maybe we don't need this local variable.

cdumez replied:
I still think `return String::fromUTF8(xpc_dictionary_get_string(xpcObject, "identifier"));` would look nicer.

pvollan replied:
Fixed!

cdumez commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -27,9 +27,29 @@
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BEWebContentProcess (ExtensionProcess)
++(void)webContentProcessWithBundleID:(NSString* _Nullable)bundleID interruptionHandler:(void(^)(void) _Nullable)interruptionHandler completion:(void(^)(BEWebContentProcess* _Nullable process, NSError* _Nullable error))completion;
```
Stars on wrong side.

cdumez replied:
Starts are still on the wrong side.

pvollan replied:
Done!

cdumez commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -27,9 +27,29 @@
+@end
+
+@interface BENetworkingProcess (ExtensionProcess)
++(void)networkProcessWithBundleID:(NSString* _Nullable)bundleID interruptionHandler:(void(^)(void) _Nullable)interruptionHandler completion:(void(^)(BENetworkingProcess* _Nullable process, NSError* _Nullable error))completion;
```
ditto.

cdumez replied:
Stars are still on the wrong side.

cdumez commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -27,9 +27,29 @@
+@end
+
+@interface BERenderingProcess (ExtensionProcess)
++(void)renderingProcessWithBundleID:(NSString* _Nullable)bundleID interruptionHandler:(void(^)(void) _Nullable)interruptionHandler completion:(void(^)(BERenderingProcess* _Nullable process, NSError* _Nullable error))completion;
```
ditto.

cdumez replied:
Starts are still on the wrong side.

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -51,24 +52,19 @@
 
 #if USE(EXTENSIONKIT)
-static RetainPtr<_SEGrant> grantCapability(_SECapability *capability, _SEExtensionProcess *process)
+static RetainPtr<BEProcessCapabilityGrant> grantCapability(BEProcessCapability *capability, std::optional<ExtensionProcess> process)
```
Should pass the `process` as a `const std::optional<ExtensionProcess>&` to avoid copying.

cdumez replied:
Feedback not taken into account.

pvollan replied:
Ah, good point, will fix!

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -255,9 +251,16 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+    invokeAsync(granterQueue(), [platformCapability = capability.platformCapability(), platformMediaEnvironment = capability.platformMediaEnvironment(), isActive] {
 #if USE(EXTENSIONKIT)
-        if ([platformCapability setActive:isActive])
+        NSError* error = nil;
```
Star on wrong side.

cdumez replied:
Star still on the wrong side.

pvollan replied:
Ah, good catch, will fix!

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -124,7 +127,7 @@ class ProcessLauncher : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<P
 
 #if USE(EXTENSIONKIT)
-    RetainPtr<_SEExtensionProcess> extensionProcess() const { return m_process; }
+    std::optional<ExtensionProcess> extensionProcess() const { return m_process; }
```
should return a `const std::optional<ExtensionProcess>&` to avoid copying.

cdumez replied:
Still not taken into account.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessExtension.h:
```diff
@@ -0,0 +1,63 @@
+OBJC_PROTOCOL(BEProcessCapabilityGrant);
+OBJC_PROTOCOL(UIInteraction);
+
+using ExtensionProcessVariant = std::variant<std::monostate, RetainPtr<BEWebContentProcess>, RetainPtr<BENetworkingProcess>, RetainPtr<BERenderingProcess>>;
```
What is the `std::monostate, ` used for? It seems you're passing a std::optional of this variant around already? Do we really need this extra state vs std::nullopt?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessExtension.h:
```diff
@@ -0,0 +1,63 @@
+
+class ExtensionProcess {
+public:
+    ExtensionProcess() { }
```
`= default;`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessExtension.h:
```diff
@@ -0,0 +1,63 @@
+    ExtensionProcessVariant m_process;
+};
+
+}
```
`// namespace WebKit`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -155,14 +152,14 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
 
 #if USE(EXTENSIONKIT)
-    auto handler = [](ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, _SEExtensionProcess* process, ASCIILiteral name, NSError* error)
+    auto handler = [](ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, ExtensionProcess process, ASCIILiteral name, NSError* error)
```
Shouldn't pass `ExtensionProcess` by value.

cdumez replied:
Star on wrong side.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -155,14 +152,14 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
         }
         if (error) {
-            NSLog(@"Error launching process %@ error %@", process, error);
+            NSLog(@"Error launching process: %@", error);
```
Do we really need this in addition to the RELEASE_LOG_FAULT ?

pvollan replied:
That is a good point, I have removed this in the latest patch, along with addressing the other issues.

Thanks for reviewing!

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -28,19 +28,22 @@
 protected:
     ExtensionCapability() = default;
+
+    RetainPtr<BEProcessCapability> m_platformCapability;
```
This should be private. protected data members are not a good design.

pvollan replied:
Ah, good point, fixed!

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,112 @@
+        [process invalidate];
+    }, [&] (RetainPtr<BERenderingProcess> process) {
+        [process invalidate];
+    });
```
Shouldn't we set `m_process` to its monostate then?

pvollan replied:
That is a very good point. I think this behavior is matching the previous behavior, where we would only invalidate the process object here, but deallocate the object when the ProcessLauncher is deleted.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,112 @@
+        interaction = [process createVisibilityPropagationInteraction];
+    });
+    return interaction;
+
```
Unnecessary blank line.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,112 @@
+
+}
+
+}
```
`// namespace WebKit`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
     case ProcessLauncher::ProcessType::Web: {
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEContentProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BEWebContentProcess* _Nullable process, NSError* _Nullable error) {
```
Stars on wrong side.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEContentProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BEWebContentProcess* _Nullable process, NSError* _Nullable error) {
+            auto extensionProcess = ExtensionProcess(process);
```
`ExtensionProcess extensionProcess(process);` or `ExtensionProcess extensionProcess { process };`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
     case ProcessLauncher::ProcessType::Network: {
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SENetworkProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BENetworkingProcess* _Nullable process, NSError* _Nullable error) {
```
Stars on wrong side.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SENetworkProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BENetworkingProcess* _Nullable process, NSError* _Nullable error) {
+            auto extensionProcess = ExtensionProcess(process);
```
`ExtensionProcess extensionProcess(process);` or `ExtensionProcess extensionProcess { process };`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
     case ProcessLauncher::ProcessType::GPU: {
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEGPUProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BERenderingProcess* _Nullable process, NSError* _Nullable error) {
```
Stars on wrong side.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -84,33 +83,33 @@
-        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEGPUProcess* _Nullable process, NSError* _Nullable error) {
-            handler(weakProcessLauncher, process, name, error);
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](BERenderingProcess* _Nullable process, NSError* _Nullable error) {
+            auto extensionProcess = ExtensionProcess(process);
```
`ExtensionProcess extensionProcess(process);` or `ExtensionProcess extensionProcess { process };`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -157,14 +157,13 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
 
 #if USE(EXTENSIONKIT)
-    auto handler = [](ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, _SEExtensionProcess* process, ASCIILiteral name, NSError* error)
+    auto handler = [](ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, ExtensionProcess& process, ASCIILiteral name, NSError *error)
```
Given that you're doing a WTFMove() of the process in the implementation. I think we should be honest here and use `ExtensionProcess&&`. Otherwise it may be super confusing to the caller to get its variable moved out.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -28,19 +28,24 @@
     virtual String environmentIdentifier() const = 0;
-    virtual RetainPtr<_SECapability> platformCapability() const = 0;
+    BEProcessCapability *platformCapability() const { return m_platformCapability.get(); }
+    void setPlatformCapability(BEProcessCapability *capability) { m_platformCapability = capability; }
```
Why not pass it to the constructor instead?

cdumez replied:
A setter only makes sense if you need to update the value later on, after construction. However, it doesn't seem you do so passing it to the constructor would make more sense.

pvollan replied:
That is a good point. I tried doing this, and in the case when we're initializing MediaCapability, we need to initialize m_mediaEnvironment first, which may require us to have a setter. Given this, would you be ok with keeping the setter?

cdumez replied:
Ok but please mark it as `protected` instead of `public` at least.

pvollan replied:
Ah, good call, thank you!

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,111 @@
+
+void ExtensionProcess::invalidate() const
+{
+    WTF::switchOn(m_process, [&] (std::monostate) {
```
Why is this still dealing with `std::monostate`?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,111 @@
+{
+    NSError *error = nil;
+    OSObjectPtr<xpc_connection_t> xpcConnection;
+    WTF::switchOn(m_process, [&] (std::monostate) {
```
Why still dealing with monostate?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,111 @@
+{
+    NSError *error = nil;
+    RetainPtr<BEProcessCapabilityGrant> grant;
+    WTF::switchOn(m_process, [&] (std::monostate) {
```
Why still dealing with monostate?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
```diff
@@ -0,0 +1,61 @@
+
+class ExtensionProcess {
+public:
+    ExtensionProcess() = default;
```
Does this really need a default constructor? It will construct a ExtensionProcessVariant with a null `RetainPtr<BEWebContentProcess>`, which doesn't seem terribly useful.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
```diff
@@ -0,0 +1,61 @@
+
+} // namespace WebKit
+
+#endif
```
`// USE(EXTENSIONKIT)`

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2970,7 +2970,7 @@ void NetworkProcess::setIsHoldingLockedFiles(bool isHoldingLockedFiles)
         m_holdingLockedFileAssertion = nullptr;
 #if USE(EXTENSIONKIT)
-        invalidateGrant();
+        invalidateActivity();
```
This name is too generic. I assume you mean `invalidateFileActivity()`

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -2979,9 +2979,9 @@ void NetworkProcess::setIsHoldingLockedFiles(bool isHoldingLockedFiles)
 
 #if USE(EXTENSIONKIT)
-    if (hasAcquiredGrant())
+    if (hasAcquiredActivity())
```
This name is too generic. We should probably use `hasAcquiredFileActivity`

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -524,9 +524,9 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
-    bool acquireLockedFileGrant();
-    void invalidateGrant();
-    bool hasAcquiredGrant() const;
+    bool acquireLockedFileActivity();
```
Interestingly, this one is correctly named but the others are not.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -0,0 +1,107 @@
+{
+    NSError *error = nil;
+    OSObjectPtr<xpc_connection_t> xpcConnection;
+    WTF::switchOn(m_process, [&] (RetainPtr<BEWebContentProcess> process) {
```
Would you mind trying to the following to see if it works? If it does, please fix all `switchOn()`  calls in this file accordingly:
```
WTF::switchOn(m_process, [&](auto& process) {
    xpcConnection = [process makeLibXPCConnectionError:&error];
});
```

I think a single visitor with `auto` should in theory suffice and would simplify this class implementation a lot.

pvollan replied:
That works great, thank you!

cdumez reviewed (approved):
LGTM with nits.

pvollan committed:
Adopt API for process launch and capabilities
https://bugs.webkit.org/show_bug.cgi?id=268644
rdar://122196452

Reviewed by Brent Fulgham and Chris Dumez.

Adopt BrowserEngineKit API for process launch and capabilities.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setIsHoldingLockedFiles):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkProcessCocoa.mm:
(WebKit::NetworkProcess::acquireLockedFileActivity):
(WebKit::NetworkProcess::invalidateActivity):
(WebKit::NetworkProcess::hasAcquiredActivity const):
(WebKit::NetworkProcess::acquireLockedFileGrant): Deleted.
(WebKit::NetworkProcess::invalidateGrant): Deleted.
(WebKit::NetworkProcess::hasAcquiredGrant const): Deleted.
* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::platformCapability const):
* Source/WebKit/Platform/cocoa/ExtensionCapability.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ExtensionCapabilityGrant::ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::setPlatformGrant):
* Source/WebKit/Platform/cocoa/MediaCapability.h:
* Source/WebKit/Platform/cocoa/MediaCapability.mm:
(WebKit::createPlatformCapability):
(WebKit::MediaCapability::environmentIdentifier const):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::extensionProcess const):
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::grantCapability):
(WebKit::grantCapabilityInternal):
(WebKit::ExtensionCapabilityGranter::setMediaCapabilityActive):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::gpuProcessCachesDirectory):
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::extensionProcess const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::launchWithExtensionKit):
(WebKit::invalidateProcess):
(WebKit::ProcessLauncher::platformDestroy):
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::platformInvalidate):
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]):
(-[WKContentView _setupVisibilityPropagationViewForGPUProcess]):

Canonical link: https://commits.webkit.org/274822@main
